kratos/python_scripts/python_linear_solver_factory.py

```diff
@@ -0,0 +1,39 @@
+from __future__ import print_function, absolute_import, division # makes KratosMultiphysics backward compatible with python 2.6 and 2.7
+
+def ConstructSolver(configuration):
+    import KratosMultiphysics as KM
+
+    if(type(configuration) != KM.Parameters):
+        raise Exception("input is expected to be provided as a Kratos Parameters object")
+
+    solver_type = configuration["solver_type"].GetString()
+
+    linear_solver_apps = {
```

Comment:
Why the list is specified manually?, the KratosComponents has a register

Reply:
Things are not added to the KratosComponents register unless you import the relevant application first. How do you know which application you need to import? @philbucher suggests with a list (at least until something better is implemented).

Reply:
Ok then, right now I cannot think a better solution

Reply:
Is there any way to put these lists in the `__init__` of these modules and retrieve them in python before importing? Just to avoid being hard coded?

Reply:
@pooyan-dadvand which `__init__` do you mean?

The problem is that somewhere we have to say that some solvers are defined in other apps
Only then we can import these apps, otherwise how would we know which app to import?

Reply:
Check e.g. the `linear_solver_factory.py`
There if e.g. `PastixIterative` is selected then the ExternalSolversApp is imported

Reply:
In the `__init__` of external solver application and other external solvers. My idea was to have a list of "external applications" in factory and just go to their module and ask for their registered solvers before importing them in a try catch:

```python
try:
  import ExternalSolvers.RegisteredSolvers
  if ExternalSolvers.RegisteredSolvers.Has(solver)
     import ...
```
But I don't know if is possible

Reply:
ok now I see

but how would you do `import ExternalSolvers.RegisteredSolvers` without importing the ExternalSolversApp?

Reply:
ping @pooyan-dadvand 
this is kinda urgent since the linear solvers are not working through the factory

Reply:
wouldn't it be enough to add ExternalSolversApplication in the main script and everything works? (i mean, without this fix)

Reply:
As a short-term solution this fixes the problem, this is in fact what I am telling ppl to do when they ask how to fix it (=> talking abt `MainKratos.py`, NOT abt the `StructuralMechanicsAnalysis.py`)

In the long term however we need a better solution (aka this PR) for several reasons:
1. What abt the Linear-Solvers from the EigenSolversApp? => this is actually where the "good" (aka pardiso) direct- solvers are
2. The ExternalSolversApp gives problems sometime so we cannot import it by default (e.g. windows, clusters)
=> Also conditionally importing Apps (with `try catch`) is not always working, I recently struggled a lot with this
3. What if in the future another app will be created that also contains linear-solvers? E.g. sth GPU-based. It should be possible to only change the name of the solver in the ProjectParams and not by importing the app

Reply:
> My feeling is that ExternalSolversApplication or EigenSolversApplication ar not really separate applications but rather modules which may be available or not.
> 
> I just wonder if we should not include them as a optional part of kratos, so that if they are available they are automatically loaded when kratos is loaded.

I agree, but I would not load them by default even if they are available. As I said I recently had problems in the Cluster with only loading the ExternalSolversApp. Also on Win I think it can be problematic. 



> Also my feeling is that a cleaner solution would be to be able to tell which applications we specifically want to be loaded in main script rather than this approach.
> 
> I am not 100% sure though...

I think since now it is no longer necessary (we discussed this recently and removed the checks) to load the apps in the MainKratos, it is a nice feature that they can only be loaded if needed (taking as example ShapeOptimization where only sometimes meshmoving is needed)

=> I think it makes it much easier for the user if one does not have to know all the implementational details. We ofc know this but imagine a user trying to use a more advanced linear solver (e.g. pardiso) then how would they know that this can only be done after loading a certain app...? => then we get Issues in github like "Available linear solvers are all slow"

What is your main problem with this? I know it is not the best solution but it does the job until we have something better. Plus also this will not change much as not every week someone adds a linear solver